### PR TITLE
eth/catalyst: build the execution payload async

### PR DESCRIPTION
**Pull request by MariusVanDerWijden**

### CODE REVIEW

**Review by djrtwo — APPROVED** [review]

**Comment by karalabe** [comment]
I'm unsure about this. There's no block reward post PoS, so creating an empty block is worse than not creating a block at all.

If the goal is to not hang the API, we should just start an async block producer and pull the block (or block) when needed. Also IMHO this should be done via the miner, not in the catalyst API.

Also starting a goroutine and never caring if it even terminates is a recipe for a OOM (someone feeds us newPayloads in a tight loop and we keep minding blocks in a background thread).

TL;DR: We should notify the miner to make us a new block when needed, return a "stub" that's useful for the beacon response id but not yet filled, and when actually requested, block before returning (maybe with a timeout). If a new mining request is filed, abort the old one.

**Comment by MariusVanDerWijden** [comment]
Alright, I moved the async block production to the miner package now. 
There are two "ugly" things:
- I had to export the GetWorkReq type. I've tried multiple other methods, but they all resulted in us not conforming to the spec wrt. payloadIDs and being able to answer old payload requests
- RequestSealingBlock runs an additional `prepareWork` in order to verify some preconditions. The result of this operation is discarded. However without this check we can not conform to the spec either (return error on invalid payload parameters). I also tried to reuse the environment, but it is way too complicated code for very little gain (also discarding the prefetcher becomes a major issue)

**Review by rjl493456442 — APPROVED** [review]
LGTM, but please rebase it.

File: eth/catalyst/queue.go [hunk 1]
```diff
@@ -34,11 +35,47 @@ const maxTrackedPayloads = 10
...
+			// TODO(rjl49345642, Marius), should we keep this
+			// 100ms timeout allowance? Why not just use the
+			// default and then fallback to empty directly?
+		}
+	}
+	block := req.empty
+	if req.block != nil {
+		block = req.block
+	}
+	return beacon.BlockToExecutableData(block)
```
Comment (karalabe) [inline eth/catalyst/queue.go:71]: This feels a bit funky :)

```go
	block := req.empty
	if req.block != nil {
		block = req.block
	}
	return beacon.BlockToExecutableData(block)
```

Why not:

```go
	if req.block != nil {
		return beacon.BlockToExecutableData(req.block)
	}
	return beacon.BlockToExecutableData(req.empty)
```

File: eth/catalyst/queue.go [hunk 2]
```diff
@@ -34,11 +35,47 @@ const maxTrackedPayloads = 10
...
+	// Try to resolve the full block first if it's not obtained
+	// yet. The returned block can be nil if the generation fails.
+	if !req.done {
+		select {
+		case req.block = <-req.result:
+			req.done = true
+		case <-time.NewTimer(time.Millisecond * 100).C:
+			// TODO(rjl49345642, Marius), should we keep this
+			// 100ms timeout allowance? Why not just use the
+			// default and then fallback to empty directly?
```
[the remark this answers is not in the document]
Comment (karalabe) [reply eth/catalyst/queue.go:64]: Ah, also, new time used like this will leak goroutines. You should do it prior to select:

```
timeout := time.NewTimer(500 * time.Millisecond)
defer timeout.Stop()

select {
...
}
```

[the remark this answers is not in the document]
Comment (karalabe) [reply eth/catalyst/queue.go:64]: Also always use the number first and then `* time.Millisecond`. The reverse way used above reads weird.

File: eth/catalyst/queue.go [hunk 3]
```diff
@@ -34,11 +35,47 @@ const maxTrackedPayloads = 10
 // latest one; but have a slight wiggle room for non-ideal conditions.
 const maxTrackedHeaders = 10
 
+// payload wraps the execution payload generated by miner
```
Comment (karalabe) [inline eth/catalyst/queue.go:38]: Perhaps expanding it would make the doc a bit more useful:

```
// payload wraps the miner's block production channel, allowing the mined block
// to be retrieved later upon the GetPayload engine API call.
```

File: eth/catalyst/queue.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (MariusVanDerWijden) [reply eth/catalyst/queue.go:64]: Thats the asian way, from left to right

File: miner/worker.go [hunk 4]
```diff
@@ -1177,7 +1178,13 @@ func (w *worker) commit(env *environment, interval func(), update bool, start ti
 }
 
 // getSealingBlock generates the sealing block based on the given parameters.
-func (w *worker) getSealingBlock(parent common.Hash, timestamp uint64, coinbase common.Address, random common.Hash) (*types.Block, error) {
+// The generation result will be passed back via the given channel no matter
+// the generation itself succeeds or not.
+func (w *worker) getSealingBlock(parent common.Hash, timestamp uint64, coinbase common.Address, random common.Hash, noTxs bool) (chan *types.Block, chan error, error) {
+	var (
+		resChan = make(chan *types.Block, 1)
```
Comment (karalabe) [inline miner/worker.go:1185]: General nitpick, channel used in the code are usually abbreviated to `Ch`, not `Chan` :P Pls fix throughout
